### PR TITLE
ua builder suggestion

### PR DESCRIPTION
**Pull request by mickeyn**
this is just an example of how I would approach the suggested changes with a ua object builder... these changes are untested and bugs are likely to be present :)

### CODE REVIEW

**Comment by mickeyn** [comment]
the BUILDARGS in `PONAPI::Client::UA` is also ugly and we should probably find a nicer way to deal with that (can be made easy if you want to pass the type as a separate arg from ua_args to the client's constructor)

File: Client/lib/PONAPI/Client/UA.pm [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    return \%args;
+}
+
+
+sub build {
+    my $self = shift;
+
+    return (
+        $self->type eq 'Hijk' ? PONAPI::Client::UA::Hijk->new( @{ $self->ua_args } ) :
```
Comment (ericherman) [inline Client/lib/PONAPI/Client/UA.pm]: Hard-coding classes? Surely, we want decoupled dependency injection.
Shouldn't the caller pass in the pre-constructed UA?

Comment (mickeyn) [reply Client/lib/PONAPI/Client/UA.pm]: yeah, as said, I don't like this BUILDARGS at all... this PR is more of a draft and I would like to remove this part.

This whole part has nothing to do with caller passing an object - this is only called in the builder, so only in the case such object wasn't passed.
This solution is attempting to also support the ability of not passing an object but just declaring the type of UA and let the constructor deal with it. 

to simplify it, I can separate `ua_args` from `ua_type` to support this feature with the BUILDARGS ugliness.

**Comment by mickeyn** [comment]
@ericherman I've simplified my suggestion... hope this is better.

**Comment by ericherman** [comment]
I'm on holiday right now, sorry for the late reply.

I still really do not think it is good design to have knowledge of all subclasses in one place like can be seen in PONAPI::Client::UA->build ....

Rather than have "ua_type" which is essentailly an enum used to index into switch-case-like list of constructors (so far only 2), I would far rather have a function pionter to the constructor itself. Or perhaps a class name or reference which would allow for something like: $self->ua_class->new($self->ua_args); ... with these sorts of approaches people are enabled to write their own instances of the interface without having to also alter the PONAPI::Client::UA itself.

Does that make sense?

**Comment by mickeyn** [comment]
keep in mind this is not excluding the ability to pass a ready-made object, it's just an extra ability.
if you think it's totally redundant, it can be removed.

this was a draft of an idea, not a proper solution.

I am away as well... so let's wait until we're both back and sort this out in person.

**Comment by mickeyn** [comment]
ok, I removed the type/args part... kept the role to enforce an interface and the ua attribute as ro

does it make sense now?

**Comment by atomicstack** [comment]
Cherry-picked + tweaked the commit a little (and caused a merge conflict), so best to pull again from trunk (sorry!).

**Pull request closed without merging**